experiments/sharedUtils/BoxGenerator.js

```diff
@@ -1,127 +1,157 @@
+    this.numButtons = numButtons
 }
 
 //generate the box. Takes in a config (what turns on what) and returns a function that
 // determine what is turned on
 function generateBox(config){
     //function to return, which will take in a setting and return what is turned on
+    function configDict(){
+        let dict = {} // keys: setting of buttons, values: lights
+        for(var i = 0; i < (Math.pow(2, config.numButtons)); i++){
+            var set = ""
+            var setting = {}
+            for(var j = 1; j <= config.numButtons; j++){
+                if((i%(Math.pow(2,j)) >= (Math.pow(2, j-1)))){
+                    set = set + "1"
+                }
+                else {
+                    set = set + "0"
+                }
+                setting[`button${j}`] = (i%(Math.pow(2,j)) >= (Math.pow(2, j-1)))
+            }
+            dict[set] = box(setting)
+        }
+        return dict
+    }
     function box(setting){    
         let arr = [] //array of results
-        for(let i = 0; i < config.controls.length; i++){
-            arr.push(config.controls[i].eval(setting.args))
+        for(var i = 0; i < config.controls.length; i++){
+            arr.push(config.controls[i].eval(setting))
         }
         return arr //return the array
     }
-    return box //return the function
+    return configDict //return the function
 }
 
 // Takes in a configuration, and returns a string representation
 function toString(config){
     var dict = {}
-    for(let i = 0; i < config.controls.length; i++){
+    for(var i = 0; i < config.controls.length; i++){
         dict[`Light ${i+1}`] = {
-            type: config.controls[i].type
-        }
-        for(let j = 0; j < config.controls[i].arg.length; j++){
-            dict[`Light ${i+1}`][`b_${j+1}`] = config.controls[i].arg[j]
+            type: config.controls[i].type,
+            controls: config.controls[i].arg,
         }
     }
     return JSON.stringify(dict)
 }
 
-function createRandomBox(numButtons, numLights){
+function createRandomBox(numButtons, numLights, numStars){
     let configArray = []
     // create a control for each light
     for(let i = 0; i < numLights; i++){
-        console.log("creating a control for light: " + i)
-        let buttonArray = []
-        for(let j = 0; j < numButtons; j++){
-            buttonArray.push(Math.floor(Math.random()*2) === 1)
-        }
+        let compsArray = [] //array of components that control the light
+        let compsDict = {} //dict of components and their settings
+        do{
+            compsArray = []
+            compsDict = {}
+            for(let j = 0; j < numButtons; j++){
+                if(Math.floor(Math.random()*2) === 1){
+                    compsDict[`button${j+1}`] = (Math.floor(Math.random()*2) === 1)
+                    compsArray.push(`button${j+1}`)
+                }
+            }
+        } while(compsArray.length < 2 && numButtons >= 2); //keep going until there is at least one control
```

Comment:
My understanding of this is: sample a few buttons and true/false values be the buttons in charge of a light, *unless* we sample a singular rule, in which case we just replace compsArray with a singular array (line 110) - is that correct?

But can't `compsArray` be length 1, in which case the `AndObj` and `OrObj`s below only have one rule?

Reply:
Just a few commentst:

1. Does the `numButtons >= 2` matter in this `while` loop? It doesn't seem like you're modifying that in the loop, so won't it always be true as long as the config has at least 2 buttons?
2. Can you switch to just a standard `while` loop? The `do...while` construction is a bit rare to see nowadays, and I don't think it matters (compsArray will start out with 0 length so we will definitely enter the while loop)

In fact I would structure it as follows

```
while compsArray has less or equal to 2 elements
    randomly select button; randomly sample true/false
    if button not in compsArray
        add button to compsArray
```
this seems conceptually more simple than this do/while with a nested for loop - let me know if it makes sense